oci.jq

```diff
@@ -87,3 +87,52 @@ def sort_manifests:
 	sort_by(.platform | sort_split_platform)
 	| sort_attestations
 ;
+
+def validate_media_type:
+    if .mediaType != "application/vnd.oci.image.index.v1+json" and .mediaType then # TODO drop the second half of this validation: https://github.com/moby/buildkit/issues/4595
+        error("unsupported index mediaType: " + .mediaType)
+    else . end
+;
+
+def validate_index_type:
+    if .schemaVersion != 2 then
+        error("unsupported schemaVersion: " + .schemaVersion)
+    else . end
+    | validate_media_type
+;
+
+# jq -s 'include "oci"; validate_oci_basic' dir/oci-layout dir/index.json 
+def validate_oci_basic:
+    if length != 2 then
+        error("Unexpected oci-layout. Expecting 'oci-layout' and 'index.json'")
+    else . end
+    | .[0] |= (
+        if .imageLayoutVersion != "1.0.0" then
+            error("unsupported imageLayoutVersion: " + .imageLayoutVersion)
+        else . end
+    )
+    | .[1] |= (
+        . | validate_index_type
+        | if .manifests | length != 1 then
+            error("expected only one manifests entry, not " + (.manifests | length | tostring))
+        else . end
+
+        | .manifests[0] |= (
+            validate_media_type
+            # TODO validate .digest somehow (`crane validate`?) - would also be good to validate all descriptors recursively
+            | if .size < 0 then
+                error("invalid descriptor size: " + .size)
+            else . end
+        )
+    )
+;
+
+def image_digest($os; $arch):
+  if length != 1 then
+    error("unexpected image index document count: " + (length | tostring))
+  else .[0] end
+  | validate_index_type
+  | .manifests[]
+  | select(.platform.os == $os and .platform.architecture == $arch)
+  | .digest
+;
```

Comment:
What's the use case for this `image_digest` function?  It seems like it's dual purpose "validate the underlying image index" _and_ "extract the (platform specific) image manifest digest" ?

I think in the case of buildkit/buildx output, we can be both more specific _and_ more prescriptive here. :thinking:

ie, I'm thinking we can validate that the index itself contains at least 1 but at most 2 entries, that the second one specifically is our expected provenance object (via annotations / platform), and then we can safely assume the first _must_ be the image (perhaps with some basic media type / size validation). :thinking:

(those are all assumptions we have, but that aren't currently validated anywhere and _should_ be OK if they're broken, but if we're assuming them anyhow, it's valuable to validate them, especially if we're parsing the file anyhow)

Reply:
Given the scale of the change for signing, I personally think that would be premature optimization and can easily be refined in a subsequent PR without affecting the function signature.

Reply:
@LaurentGoderre can you expand on how adding more sanity checks would be a premature optimization? After the SBOM incident, it seems prudent to assert on as much as we know.

Reply:
ok, how about:

```
def image_digest($os; $arch):
    if length != 1 then
        error("unexpected image index document count: \(length)")
    else .[0] end
    | validate_oci_index
    .manifests
    | (
        map(
            select(.annotations["vnd.docker.reference.type"] == "attestation-manifest")
        )
        | if length != 1 then
            error("attestation manifest not found")
        else .[] end
        | .annotations["vnd.docker.reference.digest"]
        | if . == null then
            error("image reference for attestation manifest not found")
        else . end
    ) as $digest
    | map(
        select(.digest == $digest)
    )
    | if length == 0 then
        error("image digest not found in index")
    else .[] end
    | if .platform.architecture != $arch then
        error("architecture does not match expected architecture \($arch)")
    else . end
    | if .platform.os != $os then
        error("os does not match expected os \($os)")
    else . end
    | $digest
;
```

Reply:
Nope, it's cool how it work but the attestation manifest is not guaranteed to be there (OCI Layout or if the manifest is a referrer)

Reply:
Maybe I'm not saying it right but I meant in case where the attestations manifest were not in the index but somewhere else (like a referrer)

Reply:
Sorry, let me try to rephrase without the added validation component so I can try to be more clear about my concerns (and then we can discuss them better). :bow: :innocent:

My main concerns with this function are the `$os` + `$arch` inputs and the generic name.

The easier of these is the naming; the name ought to be more descriptive of the purpose of the function, and that's easy to fix if we can get a better understanding of how and where it's intended to be used, which it sounds like you've got more use cases in mind than our current use case.  Is that accurate?

My concern with the `$os` and `$arch` inputs is related: it's effectively implementing a tiny subset of https://github.com/containerd/platforms without respect for all the edge cases (for example, `variant`), so as above, I feel like I need to understand better what the intended usage is to provide better/more specific suggestions for how to improve it.

As far as I can see from our current use case, this could be much simpler (and input-less) because the digest we're interested in is _always_ the first one (except in the case of `oci-import` builds, where the `index.json` itself is the "index digest" we're looking for), and if that isn't true, we have an issue and should be failing the build (hence my bringing further validation into the discussion).

Perhaps that helps? :sweat_smile:

Reply:
But isn't the order also an assumption?

Reply:
Not needed anymore since we are doing it in bash

Reply:
we are looking for the digest of the manifest that has the layers. By doing it in bash we can skip the traversing of the hierarchy and find the right one directly.

Reply:
The bash solution works for oci import just as well as as standard builds.